### PR TITLE
wagmi

### PR DESCRIPTION
**Pull request by BrickheadJohnny**
## web3-react/ethers -> wagmi/viem migration

**General**

This PR became pretty big, but I suggest you to start by reviewing these minor changes:

- renamed `account` to `address` (`useWeb3React` -> `useAccount`)
- renamed `isActive` to `isConnected` (`useWeb3React` -> `useAccount`)
- getting `chainId` from the `useChainId` hook (`useWeb3React` -> `useChainId`)
- we're using `publicClient` or `walletClient` instead of a `provider` from now (`useWeb3React` -> `usePublicClient` or `useWalletClient`)
- removed the `useReverseEnsName` hook, because wagmi provides a `useReverseENSName` hook by default, so we can use that
- wagmi types EVM addresses as `0x${string}`, so I decided to use this type in a couple of different places (so we don't get type errors, e.g. for the `pickedCurrency` state in the Guild checkout context)
- using the built-in `useToken` hook instead of `useTokenData` where possible
- removed the custom `useContract` hook & using the `useContractRead`/`useContractReads` hooks: when we needed to fetch data from another chain, we instantiated a `JsonRpcBatchProvider` & a `Contract`, then used SWR to do a contract call and cache the data. We don't need to do that anymore, we can just use the built-in wagmi hooks to do a contract read call on any supported chain.
- the "switch network" functionality became a bit cleaner, since we can use the `useSwitchNetwork` hook for that too (`Web3ConnectionManager`)
- removed the `useEagerConnect` hook, we can use the built-in `autoConnect: true` setting
- removed the custom `useBalance` hook (we can use the built-in one)
- replaced the ethers specific types and utils with viem types and utils (e.g. `BigNumber` -> `bigint`, `parseUnits`, `formatUnits`, ...)

**Refactored chains config**

Wagmi provides many chain configurations by default (with general chain data, public RPCs, native currency data, block explorer URLs, etc.), so I refactored `connectors.ts` in the following way:

- made a `chains.ts` file, where we export the following data:
  - a `CHAIN_CONFIG` object, which contains chain configurations in wagmi's format (so we can simply re-use the built in chains from wagmi, but we can easily define custom chains too)
  - an enum with the chain names + ids (e.g. `ETHEREUM = 1`)
  - the `Chain` type
  - `chainIconUrls`, `coinIconUrls`, `blockExplorerIcons`, and `etherscanApiUrls` (these are separate objects now, it's much cleaner this way)
- `connectors.ts` now only contains the actual connector configurations

To better understand the naming conventions, I suggest you to take a look at the official [migration guide](https://viem.sh/docs/ethers-migration.html).

**New `ClientOnly` component**

I added a `ClientOnly` component to avoid hydration issues, it'll be fixed in wagmi v2 ([source](https://twitter.com/_jxom/status/1714740525142806610)), so we'll be able to remove it soon hopefully.

**New hooks**

- `useEstimateGas` - a wrapper hook for the viem `estimateContractGas` function & we're also converting the gas price to USD inside this hook
- `useSubmitTransaction` - this hook wraps the `usePrepareContractWrite`, `useEstimateGas`, `useContractWrite` & `useWaitForTransaction` hooks, because we'd use all these hooks during a contract write transaction. I haven't used this one for the NFT/Guild Pin flows, because those are a bit different (we need to call our API & get the parameter which we pass to the contract call, and then start the actual transaction - I didn't want to do this with `useEffect`s, so I used simple viem function calls for these flows, it's much cleaner this way)

**CollectNftContext fix**

I noticed that the transaction modal wasn't included in this context, so I added it in this PR.

**Final notes**

I realized that we could use a multicall inside `useUsersGuildPins` & we could also get rid of the custom `useTokenData` hook later, so I made 2 issues in Linear. These changes weren't critical for this PR, so I haven't implemented them.

### CODE REVIEW

File: src/hooks/useSubmitTransaction.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+const useSubmitTransaction = <
+  TAbi extends Abi | readonly unknown[],
+  TFunctionName extends string,
+  TChainId extends number
+>(
+  contractCallConfig: UsePrepareContractWriteConfig<TAbi, TFunctionName, TChainId>,
+  options?: {
+    setContext?: boolean
+    customErrorsMap?: Record<string, string> // TODO: maybe we can infer custom error names from the ABI too? We could experiment with it later.
```
Comment (BrickheadJohnny) [inline src/hooks/useSubmitTransaction.ts:21]: Left TODO comments here, but these aren't crucial for this PR

File: src/components/_app/Web3ConnectionManager/components/WalletSelectorModal/WalletSelectorModal.tsx [hunk 2]
```diff
@@ -169,23 +170,21 @@ const WalletSelectorModal = ({ isOpen, onClose, onOpen }: Props): JSX.Element =>
...
-                Sign message to verify that you're the owner of this account.
+                Sign message to verify that you're the owner of this address.
               </Text>
             )}
             <Stack spacing="0">
-              {connectors.map(([conn, connectorHooks], i) => {
-                if (!conn || !connectorHooks) return null
-                if (conn instanceof GnosisSafe && !conn?.sdk) return null
-
+              {connectors.map((conn) => {
```
[the remark this answers is not in the document]
Comment (cs-balazs) [reply src/components/_app/Web3ConnectionManager/components/WalletSelectorModal/WalletSelectorModal.tsx]: https://github.com/agoraxyz/guild.xyz/pull/944 should resolve this :ayy:

File: package.json [hunk 3]
```diff
@@ -24,17 +24,6 @@
...
-    "@ethersproject/bignumber": "^5.7.0",
-    "@ethersproject/bytes": "^5.7.0",
-    "@ethersproject/constants": "^5.7.0",
-    "@ethersproject/contracts": "^5.7.0",
-    "@ethersproject/hash": "^5.7.0",
-    "@ethersproject/keccak256": "^5.7.0",
-    "@ethersproject/providers": "^5.7.2",
-    "@ethersproject/strings": "^5.7.0",
-    "@ethersproject/units": "^5.7.0",
-    "@ethersproject/wallet": "^5.7.0",
```
Comment (cs-balazs) [inline package.json:37]: :feelsgoodman:

**Review by cs-balazs — COMMENTED** [review]
Awesome work! 🚀

I didn't check every payment-related change carefully, as I'm not really familiar with the original code, but from what I could tell, those changes seemed fine to me as well 💯

File: src/components/common/Layout/components/Account/components/delegate/useDelegateVaults.ts [hunk 4]
```diff
@@ -1,33 +1,80 @@
...
+    contracts: delegates
+      ?.map((delegate) =>
+        delegateContracts.map((contract) => ({ ...contract, args: [delegate] }))
+      )
+      .flat(),
+    enabled,
+  })
+
+  // WAGMI TODO: not sure if this logic is correct, we'll need to test it!
+  const results = data
```
[the remark this answers is not in the document]
Comment (cs-balazs) [reply src/components/common/Layout/components/Account/components/delegate/useDelegateVaults.ts:56]: With this change, it seemed to work! https://github.com/agoraxyz/guild.xyz/pull/935#discussion_r1377362008

File: src/hooks/useSubmit/utils/gnosisSafeSignCallback.ts [hunk 5]
```diff
@@ -1,53 +1,53 @@
...
+      functionName: "getMessageHash",
+      args: [hashMessage(message)],
+    })
 
-    /** This message gets emitted when the sign transaction is finalized, we must await it here */
+    /**
+     * This message gets emitted when the sign transaction is finalized, we must
+     * await it here
+     */
+    // WAGMI TODO: not sure if this actually works, I just followed the docs, but haven't tested it yet
```
[the remark this answers is not in the document]
Comment (cs-balazs) [reply src/hooks/useSubmit/utils/gnosisSafeSignCallback.ts:39]: I think this is not really used at the moment, sorry for not telling earlier, I just realized this 😅 
As I remember the Safe app had some problems with the WalletConnect 2.0 migration, which broke the Safe support of most of the apps through WalletConnect (this is when the Safe context-dependent solution was added)
I'm not sure if Safe works now with WalletConnect, but I'll check it out in a follow-up PR, and I'll test this part if the connection works 👍

File: src/components/common/Layout/components/Account/components/AccountModal/AccountModal.tsx [hunk 6]
```diff
@@ -123,7 +111,7 @@ const AccountModal = () => {
                       fontWeight="medium"
                       noOfLines={1}
                     >
-                      {`Connected with ${connectorName(connector)} on`}
+                      {`Connected with ${connector?.name} on`}
```
[the remark this answers is not in the document]
Comment (BrickheadJohnny) [reply src/components/common/Layout/components/Account/components/AccountModal/AccountModal.tsx]: Oh, really nice catch! Fixed in 1cdb54c8051d029a8ba51d47e69c5cc000dc445b

**Review by dovalid — APPROVED** [review]

File: src/chains.ts [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  scrollAlpha,
+} from "static/customChains"
+import { Chain as ViemChain } from "viem"
+import {
+  arbitrum,
+  arbitrumNova,
+  avalanche,
+  base,
+  baseGoerli,
+  boba,
```
Comment (Thirawat11) [inline src/chains.ts:16]: J

**Pull request merged**